src/xml2_node.cpp

```diff
@@ -37,6 +37,36 @@ extern "C" SEXP node_name(SEXP node_sxp, SEXP nsMap) {
   END_CPP
 }
 
+// [[export]]
+extern "C" SEXP nodeset_name(SEXP node_sxp, SEXP nsMap) {
+  BEGIN_CPP
+
+  int n = Rf_xlength(node_sxp);
+
+  SEXP out = PROTECT(Rf_allocVector(STRSXP, n));
+
+  for (int i = 0; i < n; ++i) {
+    SEXP node_sxp_i = VECTOR_ELT(node_sxp, i);
+
+    if (Rf_inherits(node_sxp_i, "xml_node")) {
```

Comment:
I think this would be a little clearer/safer if you made a function that returns something enum-ish that you could then `switch` on.